File: src/graph/number_of_islands.rs
use crate::Solution;

impl Solution {
    pub fn num_islands(grid: Vec<Vec<char>>) -> i32 {
        if grid.is_empty() || grid[0].is_empty() {
            return 0;
        }
        let mut ret = 0;

        let mut grid = grid;

        for i in 0..grid.len() as i32 {
            for j in 0..grid[0].len() as i32 {
                if grid[i as usize][j as usize] == '1' {
                    ret += 1;
                    Self::sink_islands(&mut grid, i, j);
                }
            }
        }

        ret
    }

    fn sink_islands(grid: &mut Vec<Vec<char>>, i: i32, j: i32) {
        // not in grid
        if i < 0 || i >= grid.len() as i32 || j < 0 || j >= grid[0].len() as i32 {
            return;
        }
        let ii = i as usize;
        let jj = j as usize;
        // already sinked
        if grid[ii][jj] == '0' {
            return;
        }

        grid[ii][jj] = '0';
        Self::sink_islands(grid, i - 1, j);
        Self::sink_islands(grid, i + 1, j);
        Self::sink_islands(grid, i, j - 1);
        Self::sink_islands(grid, i, j + 1);
    }
}
#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn test() {
        let grid = vec![
            vec!['1', '1', '1', '1', '0'],
            vec!['1', '1', '0', '1', '0'],
            vec!['1', '1', '0', '0', '0'],
            vec!['0', '0', '0', '0', '0'],
        ];
        println!("{}", Solution::num_islands(grid));
    }
}

File: src/graph/number-of-islands.py
from typing import List


class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        def sink(grid, i, j):
            if i < 0 or i >= len(grid) or j < 0 or j >= len(grid[i]):
                return
            if grid[i][j] == "0":
                return

            grid[i][j] = "0"
            sink(grid, i - 1, j)
            sink(grid, i + 1, j)
            sink(grid, i, j - 1)
            sink(grid, i, j + 1)

        if not grid:
            return 0
        ret = 0
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] == "1":
                    sink(grid, i, j)
                    ret += 1
        return ret


if __name__ == "__main__":
    s = Solution()
    grid = [
        ["1", "1", "1", "1", "0"],
        ["1", "1", "0", "1", "0"],
        ["1", "1", "0", "0", "0"],
        ["0", "0", "0", "0", "0"],
    ]
    print(s.numIslands(grid))
